### ISSUE TITLE
When using a custom reward function and the reward model is an external model, can the reward model be parallelized for generation?

### ISSUE THREAD

**Issue by Jasaxion** [volcengine/verl#1507]
When using a custom reward function and the reward model is an external model, can the reward model be parallelized for generation?
I use the GRPO method and then construct a new reward function to score the model's generation. Currently, when I train using a small modification of the example in /example/grpo_trainer, I can only call the custom reward function once each time (that is, pass the model's response to the reward model, and then the reward model returns a score), and can only generate one at a time. The reward model is deployed using vllm, which makes the generation very slow and does not make use of the concurrent advantages of vllm. Is there a way to support concurrent calls during the model training process?
Thank you very much for your help!

------

当使用自定义奖励函数，并且奖励模型为外部模型时，调用奖励模型是否可以并行生成？
我利用 GRPO 的方法，然后构建了一个新的奖励函数来对模型的生成进行评分，目前我利用少量修改后的/example/grpo_trainer中的例子进行训练时，每次只能调用一次自定义的奖励函数（即将模型的 response 传给奖励模型，然后奖励模型返回分数），只能一条一条生成，奖励模型使用 vllm 部署，这样生成速度非常慢，没有利用到 vllm 的并发优势，是否有办法支持模型训练过程中能并发调用？
初次使用 verl，感谢帮助！

**Comment by EEE1even** [comment]
> Solved by implementing batch input and output through modifying reward_manage.
> 
> 已解决，根据参考例子通过修改 reward_manage 来实现 batch 输入和输出。

可以给一个修改后的参考示例和执行code吗

**Comment by sweartoyou** [comment]
同求一个

**Comment by zhongpenggeo** [comment]
> 同求一个
在verl-workers-reward_manager-batch.py里有一个BatchRewardManager，可以实现输入是data_source、solution和gt的列表，输出是score的列表，应该符合你的需要
需要设置reward_model.reward_manager=batch, custom_reward_function.name=compute_score_=batch(这里是自定义的batch Reward函数)
```python
@register("batch")
class BatchRewardManager:
    def __init__(self, tokenizer, num_examine, compute_score, reward_fn_key="data_source", **reward_kwargs):
        self.tokenizer = tokenizer
        self.num_examine = num_examine
        self.compute_score = compute_score
        self.reward_fn_key = reward_fn_key
        self.reward_kwargs = reward_kwargs

    def verify(self, data):
        prompt_ids = data.batch["prompts"]
        response_ids = data.batch["responses"]
        attention_mask = data.batch["attention_mask"]

        prompt_len = prompt_ids.shape[-1]
        valid_response_lengths = attention_mask[:, prompt_len:].sum(dim=-1)

        responses_str = []
        for i in range(len(data)):
            valid_len = valid_response_lengths[i]
            valid_response_ids = response_ids[i][:valid_len]
            response_str = self.tokenizer.decode(valid_response_ids, skip_special_tokens=True)
            responses_str.append(response_str)

        ground_truths = [item.non_tensor_batch["reward_model"].get("ground_truth", None) for item in data]
        data_sources = data.non_tensor_batch[self.reward_fn_key]
        extras = data.non_tensor_batch.get("extra_info", [None] * len(data))

        scores = self.compute_score(
            data_sources=data_sources,
            solution_strs=responses_str,
            ground_truths=ground_truths,
            extra_infos=extras,
            **self.reward_kwargs,
        )

        return scores
```
另外一种方式是设置reward_model.launch_reward_fn_async=True可以实现对computer_score函数的并发，但我没有试过
